README.md
# More Achievements
More Achievements is a mod for factorio that adds in a large amount of achievements, ranging from simple progression based ones, to ones that will give hours of challenge, like managing to make a base that launches one rocket per minute.

It also adds achievements for certain mods when installed.

## List of all added achievements:

* Crafting manually is too slow! : Place your first assembling machine.
* Alien contact : Kill a biter for the first time.
* Bitten : Get bitten by a biter.
* The essence of Discovery! : Craft your first science pack.
* Thomas Edison : Illuminate your factory with a lamp.
* Mathematical! : Place your first arithmetic combinator.
* Friendly Fire : Destroy one of your own buildings.
* Atom Smashing : Build a nuclear reactor.
* Tango down : Have a turret kill a biter.
* Efficient miner : Reach mining productivity 16.
* Looted : Pick up an item.
* Copy and pasted : Duplicate the settings of a machine onto another.
* Accumulating : Place an accumulator.
* Depleted : Completely deplete a resource tile.
* A brisk walk : Finish the game in less than 24 hours.
* Hello world! : Talk in chat.
* Deforestation : 100 thousand trees have been destroyed on this planet.
* On the road again : Place a car.
* Fully Automatic : Upgrade your gun to a SMG.
* Unending Assault : Trigger 100 attacks by pollution.
* One RPM : Create a base that launches one rocket per minute.
* Two RPM : Create a base that launches two rockets per minute.
* Bulldozed : Destroy 500 trees by running them over.
* Let there be Light : 200 Lamps placed in the world.
* Dedication : Research an infinite tech to level 50.
* Maggots! : Upgrade your gun to a rocket launcher.
* We didn't start the fire : Upgrade your gun to a flamethrower.
* You forgot something : Launch a rocket without a satellite.
* Ghosted : Place a ghost entity.
* Praise the sun! : Have 10k solar panels placed at once.
* Blueprinted : Capture a blueprint of a setup.
* Convey your ideas : Have 10,000 transport belts placed in the world at once.
* Well balanced : Make a blueprint containing 10+ splitters.
* Gear mass production 1 : Produce 10k iron gears.
* Gear mass production 2 : Produce 1 million iron gears.
* Gear mass production 3 : Produce 20 million iron gears.
* Steampunk : Generate 10GJ of energy per hour with steam only.
* The olden days... : Win the game without building a fluid wagon.
* Trains are too hard! : Win the game without building a rail system.
* Don't panic : Launch your car into space.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.5.0
Date: 2019-6-25
  Features:
    - Mod updated to 0.17 by Gangsir.
  Removed:
    - Settings to enable and disable external mod support removed.
    - Support for Bob's mods temporarily removed. These will be returning at a later time, once I can look into how Bob updated his mods to 0.17. For now, I just want the mod to work with 0.17.
  Changes:
    - Fixed the "Efficient miner" achievement to be more achievable after the changes to the tech's scaling.
    - Changelog now uses the ingame system.
  Added:
    - Added "Apprentice Slayer" achievement.
    - Added "Slayer of worlds" achievement.
    - Added "Encampment" achievement.
    - Added "Fortress" achievement.
    - Added "Castle" achievement.
    - Added "Homewrecker" achievement.
    - Added "I'm melting" achievement.
    - Added "Shot down" achievement.
    - Added "Terraformer" achievement.
    - Added "If you build it, they will destroy it" achievement.
    - Added "Turret creeper" achievement.
    - Added "Apprentice metalsmith" achievement.
    - Added "Journeyman metalsmith" achievement.
    - Added "Master metalsmith" achievement.
    - Added "Buildin' a sentry" achievement.
    - Added "Kill it with fire" achievement.
    - Added "Laser beams" achievement.
    - Added "Wall of guns" achievement.
    - Added "Gatekeeping" achievement.
    - Added "Sticky boots" achievement.
    - Added "Drop the bass" achievement.
    - Added "Tactical nuke incoming" achievement.

---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 2019-1-1
  Features:
    - Mod updated to 0.16 by Bilka. As I wasn't maintaining the mod during this time, this entry is a placeholder to fill the gap.
  Added:
    - "Don't Panic" achievement added.

---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 2019-1-1
  Features:
    - Some minor fixes
    - Added support for both Bob's mining, and Bob's assembling machines.
    - Added settings to disable extra achievements on a per-mod basis, this includes the achievements the mod adds for vanilla.

---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 2019-1-1
  Features:
    - Some minor fixes
    - `Steampunk` achievement now requires 200GJ of energy per hour

---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 2019-1-1
  Features:
    - Added support for Bob's logistics. More mods are in the works, patch by patch.
    - A few fixes to reported bugs
    - More graphical updates

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2019-1-1
  Features:
    - Some minor fixes
    - Added a few more achievements
    - Improved several graphics, courtesy of [Bilka](https://mods.factorio.com/mods/Bilka).

---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 2019-1-1
  Features:
    - Fix depleted achievement crashing

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2019-1-1
  Features:
    - Quick hotfix

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-1-1
  Features:
    - Released the mod


vanillaScripting.lua
script.on_init(function(e)
    global.trees_destroyed = 0
end
)

function treeDestroyed(e)
    if e.entity.type == "tree" then
        global.trees_destroyed = global.trees_destroyed + 1 --increment the number of trees destroyed
        if global.trees_destroyed >= 100000 then
            for index, player in pairs(game.players) do --give the achievement to every player
                player.unlock_achievement("deforestation")
            end
        end
    end
end
script.on_event(defines.events.on_player_mined_entity, treeDestroyed)
script.on_event(defines.events.on_robot_pre_mined, treeDestroyed)

function onEntityDied(e)
    local entity = e.entity
    local cause = e.cause
    local causeForce = e.force

    --Friendly fire - destroy your own building
    if cause and causeForce and cause.type == "character" and entity.force == causeForce then
        cause.player.unlock_achievement("friendly-fire")
        --tango down - have a turret kill a biter
    elseif cause and causeForce and cause.name == "gun-turret" and entity.type == "unit" then
        for index, player in pairs(causeForce.players) do
            player.unlock_achievement("tango-down")
        end
        -- Homewrecker
    elseif cause and causeForce and cause.type == "character" and entity.type == "unit-spawner" then
        for index, player in pairs(causeForce.players) do
            player.unlock_achievement("homewrecker")
        end
        -- Turret creeper
    elseif cause and causeForce and cause.type:find("turret") and entity.type == "unit-spawner" then
        game.print("turret creeper")
        for index, player in pairs(cause.force.players) do
            player.unlock_achievement("turret-creeper")
        end
        -- Shot down
    elseif cause and causeForce and cause.type == "unit" and (entity.type == "construction-robot" or entity.type == "logistic-robot") then
        for index, player in pairs(entity.force.players) do
            player.unlock_achievement("shot-down")
        end
        -- Deforestation
    elseif entity.type == "tree" then
        treeDestroyed(e)

        -- If you build it they will destroy it
    elseif cause and causeForce and cause.type == "unit" and entity.has_flag("player-creation") then
        for index, player in pairs(entity.force.players) do
            for index, player in pairs(entity.force.players) do
                player.unlock_achievement("if-you-build-it-they-will-destroy-it")
            end
        end
    elseif not cause and entity.type:find("turret") and entity.has_flag("player-creation") then --god damn it Wube, why you gotta make turrets weird
        for index, player in pairs(entity.force.players) do
            for index, player in pairs(entity.force.players) do
                player.unlock_achievement("if-you-build-it-they-will-destroy-it")
            end
        end
    end
end
script.on_event(defines.events.on_entity_died, onEntityDied)

-- Determines if the area provided is a zero size area.
function isZeroSizeArea(area)
    return (area.left_top.x == area.right_bottom.x) and (area.left_top.y == area.right_bottom.y)
end

function onBlueprint(e)
    local player = game.players[e.player_index]
    if not player then return end
    player.unlock_achievement("blueprinted")
    if not isZeroSizeArea(e.area) and player.surface.count_entities_filtered{type = "splitter", area = e.area} >= 10 then
        player.unlock_achievement("well-balanced")
    end
end
script.on_event(defines.events.on_player_setup_blueprint, onBlueprint)

function onItemPickup(e)
    local player = game.players[e.player_index]
    if not player then return end
    player.unlock_achievement("looted")
end
script.on_event(defines.events.on_picked_up_item, onItemPickup)

-- This must be done to give the achievement to the right force that owns the drill
function onResourceDepleted(e)
    local entity = e.entity.surface.find_entities_filtered{type="mining-drill", position = e.entity.position}
    if not entity or not entity[1] then return end
    for index, player in pairs(entity[1].force.players) do
        player.unlock_achievement("depleted")
    end
end
script.on_event(defines.events.on_resource_depleted, onResourceDepleted)

function onSettingsPasted(e)
    local player = game.players[e.player_index]
    if player then
        player.unlock_achievement("copy-and-pasted")
    end
end
script.on_event(defines.events.on_entity_settings_pasted, onSettingsPasted)

function onConsoleChat(e)
    if not e.player_index then return end
    local player = game.players[e.player_index]
    if player then
        player.unlock_achievement("hello-world")
    end
end
script.on_event(defines.events.on_console_chat, onConsoleChat)

function onPlayerCrafted(e)
    local item = e.item_stack.name
    local player = game.players[e.player_index]
    if item == "submachine-gun" then
        player.unlock_achievement("fully-automatic")
    elseif item == "rocket-launcher" then
        player.unlock_achievement("maggots")
    elseif item == "flamethrower" then
        player.unlock_achievement("we-didnt-start-the-fire")
    end
end
script.on_event(defines.events.on_player_crafted_item, onPlayerCrafted)

function onPlaced(e)
    if e.created_entity.type == "entity-ghost" and game.players[e.player_index] then
        game.players[e.player_index].unlock_achievement("ghosted")
    end
end
script.on_event(defines.events.on_built_entity, onPlaced)
--script.on_event(defines.events.on_robot_built_entity, onPlaced) We don't get a player index here, so we are unable to attribute the achievement to a player

function onResearch(e)
    local research = e.research
    local force = research.force
    if research.name:find("-50") then --A level 50 research
        for index, player in pairs(force.players) do
            player.unlock_achievement("dedication")
        end
    end
end
script.on_event(defines.events.on_research_finished, onResearch)

function onRocketLaunched(event)
    local force = event.rocket.force
    -- if they launch a rocket containing no satellite and no fish
    if event.rocket.get_item_count("satellite") == 0 and event.rocket.get_item_count("raw-fish") == 0 and event.rocket.get_item_count("car") == 0 then
        for index, player in pairs(force.players) do
            player.unlock_achievement("you-forgot-something")
        end
    elseif event.rocket.get_item_count("satellite") == 0 and event.rocket.get_item_count("raw-fish") == 0 and event.rocket.get_item_count("car") >= 1 then
        for index, player in pairs(force.players) do
            player.unlock_achievement("dont-panic")
        end
    end
end
script.on_event(defines.events.on_rocket_launched, onRocketLaunched)

function onTick(e)
    if e.tick % 8000 == 0 then
        if game.surfaces[1].count_entities_filtered{name = "small-lamp"} >= 200 then
            for index, player in pairs(game.players) do
                player.unlock_achievement("let-there-be-light")
            end
        end
        if game.surfaces[1].count_entities_filtered{name = "solar-panel"} >= 10000 then
            for index, player in pairs(game.players) do
                player.unlock_achievement("praise-the-sun")
            end
        end
    end
    if e.tick % 16000 == 0 then
        if game.surfaces[1].count_entities_filtered{type = "transport-belt"} >= 10000 then
            for index, player in pairs(game.players) do
                player.unlock_achievement("convey-your-ideas")
            end
        end
    end
end
script.on_event(defines.events.on_tick, onTick)

-- I'm melting
script.on_event(defines.events.on_player_died, function(e)
    local causeEntity = e.cause
    local player = game.players[e.player_index]

    if causeEntity.name == "small-worm-turret" or causeEntity.name == "medium-worm-turret" or causeEntity.name == "big-worm-turret" then
        player.unlock_achievement("im-melting")
    end
end)

function cliffDestroyed(e)
    if e.robot then
        for i, player in pairs(e.robot.force.players) do
            player.unlock_achievement("terraformer")
        end
    else
        local player = game.players[e.player_index]
        local capsuleUsed = e.item
        if capsuleUsed.name == "cliff-explosives" then
            player.unlock_achievement("terraformer")
        end
    end
end
script.on_event({defines.events.on_player_used_capsule, defines.events.on_robot_exploded_cliff}, cliffDestroyed)
